src/ethereum/berlin/vm/instructions/environment.py

```diff
@@ -223,43 +257,24 @@ def calldatacopy(evm: Evm) -> None:
     :py:class:`~ethereum.berlin.vm.exceptions.StackUnderflowError`
         If `len(stack)` is less than `3`.
     """
-    # Converting below to Uint as though the start indices may belong to U256,
-    # the ending indices may overflow U256.
-    memory_start_index = Uint(pop(evm.stack))
-    data_start_index = Uint(pop(evm.stack))
+    # STACK
+    memory_start_index = pop(evm.stack)
+    data_start_index = pop(evm.stack)
     size = pop(evm.stack)
 
+    # GAS
     words = ceil32(Uint(size)) // 32
-    copy_gas_cost = u256_safe_multiply(
-        GAS_COPY,
-        words,
-        exception_type=OutOfGasError,
-    )
-    memory_extend_gas_cost = calculate_gas_extend_memory(
-        evm.memory, memory_start_index, size
-    )
-    total_gas_cost = u256_safe_add(
-        GAS_VERY_LOW,
-        copy_gas_cost,
-        memory_extend_gas_cost,
-        exception_type=OutOfGasError,
-    )
-    evm.gas_left = subtract_gas(evm.gas_left, total_gas_cost)
-
-    evm.pc += 1
-
-    if size == 0:
-        return
-
-    extend_memory(evm.memory, memory_start_index, size)
-
-    value = evm.message.data[data_start_index : data_start_index + size]
-    # But it is possible that data_start_index + size won't exist in evm.data
-    # in which case we need to right pad the above obtained bytes with 0.
-    value = right_pad_zero_bytes(value, size)
+    copy_gas_cost = GAS_COPY * words
+    extend_memory(evm, memory_start_index, size)
```

Comment:
My attitude is that memory extension is strictly a gas accounting phenomenon. From the EVM's point of view memory is an infinitely long buffer initialised with zeros. The only observable effect of memory expansion (to the EVM) is gas consumed.

Reply:
I would argue that since two invocations of the same opcode with the same parameters can consume different amounts of gas based on the size of memory that it is an observable effect (and that's discounting [`msize`](https://ethereum.github.io/execution-specs/autoapi/ethereum/berlin/vm/instructions/memory/index.html?highlight=msize#ethereum.berlin.vm.instructions.memory.msize), which makes it directly observable.)

I'm okay leaving as written in this PR though, if you think that's the right call.